File: MadeInJava/src/MergeSort.java
import java.util.Random;
//import java.util.Arrays; // Use this too if you want to see the array

public class MergeSort {
    public static void main(String[] args) {
        Random r = new Random();
        int[] theArray = new int[120000];

        for (int i = 0; i < theArray.length; i++)
        {// I know that just using r.nextInt() includes Integer.MAX_VALUE. Becuz of that I made it like this.
            theArray[i] = r.nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE);
        }

        long startTime = System.nanoTime();
        Sort(theArray);
        long endTime = System.nanoTime();
        //System.out.println(Arrays.toString(theArray)); // To see the array
        System.out.println(theArray.length + " integers sorted in " + ((endTime - startTime) / 1000000.0) + " milliseconds");
    }

    static void Sort(int[] numberArray)
    {
        if (numberArray.length < 2)
        {
            return;
        }

        int middle = numberArray.length / 2, i;
        int[] leftArray = new int[middle], rightArray = new int[numberArray.length - middle];

        for (i = 0; i < middle; i++)
        {
            leftArray[i] = numberArray[i];
        }

        for (i = middle; i < numberArray.length; i++)
        {
            rightArray[i - middle] = numberArray[i];
        }

        Sort(leftArray);
        Sort(rightArray);
        Merge(numberArray, leftArray, rightArray);
    }

    static void Merge(int[] numberArray, int[] leftArray, int[] rightArray)
    {
        int i = 0, j = 0, k = 0;

        while (i < leftArray.length && j < rightArray.length)
        {
            if (leftArray[i] > rightArray[j])
            {
                numberArray[k++] = rightArray[j++]; // This means numberArray[k] = rightArray[j]; k += 1; j += 1;
            }
            else
            {
                numberArray[k++] = leftArray[i++]; // This means numberArray[k] = leftArray[i]; k += 1; i += 1;
            }
        }

        while (i < leftArray.length)
        {
            numberArray[k++] = leftArray[i++]; // This means numberArray[k] = leftArray[i]; k += 1; i += 1;
        }

        while (j < rightArray.length)
        {
            numberArray[k++] = rightArray[j++]; // This means numberArray[k] = rightArray[j]; k += 1; j += 1;
        }
    }
}

File: made_in_python/merge_sort.py
from random import randint
from time import time

the_list = [0] * 12345678

for i in range(len(the_list)): # Python's integer is 64 bit. Becuz of that I made the randomizer like this.
    the_list[i] = randint(-2147483648, 2147483647 - 1) # -1 becuz the endpoint is included

def merge(number_list, left_list, right_list):
    i: int = 0
    j: int = 0
    k: int = 0

    while i < len(left_list) and j < len(right_list):
        if left_list[i] > right_list[j]:
            number_list[k] = right_list[j]
            k += 1
            j += 1
        else:
            number_list[k] = left_list[i]
            k += 1
            i += 1

    while i < len(left_list):
        number_list[k] = left_list[i]
        k += 1
        i += 1

    while j < len(right_list):
        number_list[k] = right_list[j]
        k += 1
        j += 1

def sort(number_list):
    if len(number_list) < 2:
        return

    middle: int = len(number_list) // 2
    i: int
    left_list = [0] * middle
    right_list = [0] * (len(number_list) - middle)

    for i in range(middle):
        left_list[i] = number_list[i]

    for i in range(middle, len(number_list)):
        right_list[i - middle] = number_list[i]

    sort(left_list)
    sort(right_list)
    merge(number_list, left_list, right_list)

start_time: float = time()
sort(the_list)
end_time: float = time()
#print("[" + ", ".join(str(x) for x in the_list) + "]") # To see the array
print((str)(len(the_list)) + " integers sorted in " + (str)((end_time - start_time) * 1000) + " milliseconds.")
